core/src/main/c/share/files.c

```diff
@@ -325,6 +325,52 @@ JNIEXPORT jint JNICALL Java_io_questdb_std_Files_lock
     return flock((int) fd, LOCK_EX | LOCK_NB);
 }
 
+JNIEXPORT jlong JNICALL Java_io_questdb_std_Files_openCleanRW
+        (JNIEnv *e, jclass cl, jlong lpszName, jlong size) {
+
+    jlong fd = open((const char *) lpszName, O_CREAT | O_RDWR, 0644);
+
+    if (fd < -1) {
+        // error opening / creating file
+        return fd;
+    }
+
+    jlong fileSize = Java_io_questdb_std_Files_length(e, cl, fd);
+    if (fileSize > 0) {
+        if (flock((int)fd, LOCK_EX | LOCK_NB) == 0) {
+            // truncate file to 0 byte
+            if (ftruncate(fd, 0) == 0) {
+                // allocate file to `size`
+                if (Java_io_questdb_std_Files_allocate(e, cl, fd, size) == JNI_TRUE) {
+                    // downgrade to shared lock
+                    if (flock((int) fd, LOCK_SH) == 0) {
+                        // success
+                        return fd;
+                    }
+                }
+            }
+        } else {
+            if (fileSize >= size || Java_io_questdb_std_Files_allocate(e, cl, fd, size) == JNI_TRUE) {
```

Comment:
We should perhaps be more expressive when file size is below the expected value. If this happens errorno is not set but -1 is returned. Would be good to see a test that produces actionable message

Reply:
Errno will be set in Java_io_questdb_std_Files_allocate if it cannot resize the file as far as I understand. If may look confusing but it means that if file size is greater or resize is successful return OK. So resize will be attempted when file size is below and if it fails flow falls down to unhappy path e.g., close the file and return -1.